### ISSUE TITLE
Implementing Shaved Cells in Oceananigans

### ISSUE THREAD

**Issue by siddharthabishnu** [CliMA/Oceananigans.jl#3123]
Here’s my tentative work plan for implementing shaved cells in Oceananigans:

1. Start working on the [glw-fix/partial-cells2 branch](https://github.com/CliMA/Oceananigans.jl/tree/glw/fix-partial-cells2) by @glwagner;
2. Test the existing partial cells implementation in Oceananigans. As suggested by @jm-c, compare a simulation using low vertical resolution (~10 vertical layers) and partial cells with one using high vertical resolution (~500 vertical layers) and full cells at the same horizontal resolution (to verify the effectiveness of partial cells).
3. Implement shaved cells by computing:
    - the points of intersection of the topography with the full cells;
    - the planes passing through three of these points to ‘best’ approximate the topography;
    - the portion of the area on each face (and hence the volume) of these ‘topographic’ cells belonging to the ocean, which will be necessary for diagnosing the vertical velocity.

I have gathered some references in this editable [GoogleDoc](https://docs.google.com/document/d/1UbS37g_1WeXx2UPqZo8DsmJf81ObIrWztbpZ_vNnHJE/edit?usp=sharing). Please feel free to add more.

**Comment by glwagner** [comment]
This validation test essentially implements the tracer advection test:

https://github.com/CliMA/Oceananigans.jl/blob/main/validation/immersed_boundaries/tracer_advection_over_bump.jl

but we should take care to make sure the boundary conditions are correct.

**Comment by glwagner** [comment]
@siddharthabishnu @simone-silvestri can you document the progress we discussed yesterday on this PR?

**Comment by glwagner** [comment]
Also, if there has been work do you think we can open a draft PR to track the progress?

**Comment by glwagner** [comment]
(Possibly a PR that updates the tracer advection case is all we need for now to complete the first task. I don't think the tracer advection case works on `main` currently.)

**Comment by siddharthabishnu** [comment]
@glwagner I have opened a PR with the branch sb/cut-cells-experiment on which I have been working on implementing the cut cells. To summarize the progress:

-  The two-dimensional [tracer advection test case](https://github.com/CliMA/Oceananigans.jl/blob/main/validation/immersed_boundaries/tracer_advection_over_bump.jl) over a Gaussian bump is run successfully using the low resolution from the [Adcroft](https://journals.ametsoc.org/view/journals/mwre/125/9/1520-0493_1997_125_2293_rotbsc_2.0.co_2.xml) paper and a relatively higher resolution. Both partial cells and full cells are used. Time evolution of the 'passive' tracer concentration is visualized.

- CutCellBottom is implemented in the ImmersedBoundary module similar to PartialCellBottom, and the tracer advection test case is repeated with cut cells. No noticeable difference in the results is detected so far (with the simulation using partial cells).

So, tasks 1 (a) and 2 (b) have been completed so far. Tasks 2(b) and 2(c) are being worked on right now, after completion of which tasks 1 (b) and 3 will be initiated.

**Comment by glwagner** [comment]
Nice! Do you have an image that illustrates the problems at low resolution, ie a plot that reproduces the Adcroft result?

**Comment by siddharthabishnu** [comment]
> Nice! Do you have an image that illustrates the problems at low resolution, ie a plot that reproduces the Adcroft result?

Yes. Here they are. I improved the visualization on the latest commit and pushed it right now. The types of immersed boundary and plot are included in the filenames. The discrete version of the topography (i.e. the Gaussian bump for this test case) is detectable only for the heatmaps.


https://github.com/CliMA/Oceananigans.jl/assets/12926768/f94a6b92-cb5e-4197-9e5b-bd8ea1bad42a


https://github.com/CliMA/Oceananigans.jl/assets/12926768/c2d37eb9-931e-4efa-aa99-c113e38bf393


https://github.com/CliMA/Oceananigans.jl/assets/12926768/38771b36-e030-4877-8c9f-61944cd9aefc


https://github.com/CliMA/Oceananigans.jl/assets/12926768/baf6c24b-3fc1-4f4e-aec7-17826bd4da61

**Comment by glwagner** [comment]
This paper describes a "cell merging" technique that can help reduce problems with very small cells: https://journals.ametsoc.org/view/journals/mwre/138/10/2010mwr3252.1.xml?tab_body=abstract-display

There are also some citations in there about an alternative method that they call the "thin wall" method.
